### PR TITLE
[HttpFoundation] automatically generate safe fallback filename

### PR DESCRIPTION
**Pull request by xabbuh**
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #16603 |
| License | MIT |
| Doc PR |  |

### CODE REVIEW

**Comment by Yannik** [comment]
Hi Christian,
thanks a lot for addressing this!
Maybe it's possible to convert the default filename to the nearest ascii equivalent instead of just replacing it with an underscore.
For example: 'föö.html' => 'foo.html' instead of 'f__.html', 
'fòôbàř.txt' => 'foobar.txt' instead of '______.txt'.
I am not confident in what's the best way to do this - I use https://github.com/danielstjules/Stringy#toascii for such purposes in my own applications.

Best regards, Yannik

**Comment by nicolas-grekas** [comment]
This is [RFC 2231](https://tools.ietf.org/html/rfc2231)
See also this old "paper" on the topic: http://greenbytes.de/tech/tc2231/
Doing transliterations works for western languages but is quite limited in the general case, I'm not really fan of this approach...

**Comment by xabbuh** [comment]
@nicolas-grekas Encoding a string according to RFC 2231 leads to `%` characters in the resulting string, doesn't it? This would conflict with how the `makeDisposition()` method of the `ResponseHeaderBag` is working (see https://github.com/symfony/symfony/blob/2.8/src/Symfony/Component/HttpFoundation/ResponseHeaderBag.php#L256) (but maybe we need to fix it there). What do you think?

**Comment by JHGitty** [comment]
@xabbuh Your PR is awsome! Hopefully this will be merged into Symfony.

**Comment by fabpot** [comment]
@nicolas-grekas @xabbuh Can we find something that works for everyone here? I'd like to merge this one ASAP if possible.

**Comment by xabbuh** [comment]
If we removed the limitation of not being able to have the percent character in the fallback filename, we could use RFC 2231 afaics as @nicolas-grekas proposed. But I am not sure why this check was added initially.

File: src/Symfony/Component/HttpFoundation/BinaryFileResponse.php [hunk 1]
```diff
@@ -157,6 +157,20 @@ public function setContentDisposition($disposition, $filename = '', $filenameFal
             $filename = $this->file->getFilename();
         }
 
+        if ('' === $filenameFallback && (!preg_match('/^[\x20-\x7e]*$/', $filename) || false !== strpos($filename, '%'))) {
+            $encoding = mb_detect_encoding($filename);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/BinaryFileResponse.php]: `mb_detect_encoding($string, null, true)` to enable strict mode (non-strict is useless)

File: src/Symfony/Component/HttpFoundation/BinaryFileResponse.php [hunk 2]
```diff
@@ -157,6 +157,20 @@ public function setContentDisposition($disposition, $filename = '', $filenameFal
...
+
+            for ($i = 0; $i < mb_strlen($filename, $encoding); $i++) {
+                $char = mb_substr($filename, $i, 1, $encoding);
+
+                if (ord($char) < 32 || ord($char) > 126) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/BinaryFileResponse.php]: missing handling of `%`

File: src/Symfony/Component/HttpFoundation/BinaryFileResponse.php [hunk 3]
```diff
@@ -157,6 +157,20 @@ public function setContentDisposition($disposition, $filename = '', $filenameFal
             $filename = $this->file->getFilename();
         }
 
+        if ('' === $filenameFallback && (!preg_match('/^[\x20-\x7e]*$/', $filename) || false !== strpos($filename, '%'))) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/BinaryFileResponse.php]: the if block should be moved in ResponseHeaderBag::makeDisposition IMO, isn't it?

**Comment by fabpot** [comment]
@xabbuh Do you have time to finish this one? If not, just let me know and I will finish it for you.

File: src/Symfony/Component/HttpFoundation/BinaryFileResponse.php (hunk 3 above)
Comment (xabbuh) [inline src/Symfony/Component/HttpFoundation/BinaryFileResponse.php]: The thing is that people will expect the `ResponseHeaderBag::makeDisposition()` method to throw an exception when invalid characters are passed. Silently convertin the fallback filename is a BC break imo.

**Comment by xabbuh** [comment]
@fabpot @nicolas-grekas I pushed an update.

File: src/Symfony/Component/HttpFoundation/BinaryFileResponse.php (hunk 3 above)
Comment (fabpot) [inline src/Symfony/Component/HttpFoundation/BinaryFileResponse.php]: That's a good point.

**Comment by fabpot** [comment]
:+1:

**Comment by nicolas-grekas** [comment]
:+1:

**Comment by fabpot** [comment]
Thank you @xabbuh.

**Comment by Yannik** [comment]
@fabpot Will this be merged into the more current releases of symfony?

**Comment by xabbuh** [comment]
@Yannik Yes, we regularly merge lower branches into all higher maintained branches.

**Pull request merged**